### PR TITLE
Feature: Mean Deviation Similarity Index (MDSI)

### LINKED ISSUE #113: Metric: Mean Deviation Similarity Index
**Is your feature request related to a problem? Please describe.**
It was proposed in 2017 an I missed it during initial literature review. It's performance is better that FSIMc or MS-GMSD and it may be good idea to have it in the library.

**Describe the solution you'd like**
Copy [Python implementation ](https://github.com/igv/Python-MDSI)
Or copy [original MATLAB implementation](https://www.mathworks.com/matlabcentral/fileexchange/59809)

**Describe alternatives you've considered**

**Additional context**
Link to original article: https://arxiv.org/pdf/1608.07433.pdf

image

### PR DESCRIPTION
**Pull request by denproc**
Closes #113

## Proposed Changes
- New metric `mdsi`
- New loss function `MDSILoss`
- Tests to validate aforementioned functionality
- Minor changes to other parts

### CODE REVIEW

**Comment by denproc** [comment]
@zakajd @snk4tr ready for review

File: piq/mdsi.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    else:
+        raise ValueError(f'Expected combination method "sum" or "mult", got {combination}')
+
+    mct_complex = pow_for_complex(base=gcs, exp=q).mean(dim=(-1, -2), keepdim=True)
+    score = (pow_for_complex(base=gcs, exp=q) - mct_complex).pow(2).sum(dim=0).sqrt()
+    score = (score ** rho).mean(dim=(-1, -2)) ** (o / rho)
+    if reduction == 'none':
+        return score
+    return {'mean': score.mean,
+            'sum': score.sum}[reduction](dim=0)
```
Comment (zakajd) [inline piq/mdsi.py:112]: Please check return tensor shape. All other metrics return tensor of size 0, while this metric returns tensor with size 1. 
Minimal example: 
```python
import piq
import torch
prediction = torch.rand(4, 3, 256, 256)
target = torch.rand(4, 3, 256, 256)

piq.mdsi(prediction, target).shape, piq.ssim(prediction, target).shape
```

Result: `(torch.Size([1]), torch.Size([]))`

**Comment by denproc** [comment]
@zakajd done

File: piq/functional/base.py [hunk 2]
```diff
@@ -54,3 +54,28 @@ def gradient_map(x: torch.Tensor, kernels: torch.Tensor) -> torch.Tensor:
...
+
+
+def pow_for_complex(base: torch.Tensor, exp: Union[int, float]) -> torch.Tensor:
+    r""" Power a 4D tensor with negative values or 5D tensor with complex values.
+    Complex numbers are represented by modulus and argument: r * \exp(i * \phi).
+
+    It will likely to be redundant with introduction of torch.ComplexTensor.
+
+    Args:
+        base: Tensor with shape (B x C x H x W) or (2 x B x C x H x W)
```
Comment (snk4tr) [inline piq/functional/base.py:66]: May you meant `(B x C x H x W x 2)` is case of a complex (5D) tensor?

File: piq/functional/base.py [hunk 3]
```diff
@@ -54,3 +54,28 @@ def gradient_map(x: torch.Tensor, kernels: torch.Tensor) -> torch.Tensor:
...
+        exp: int or float
```
Comment (snk4tr) [inline piq/functional/base.py:67]: Please make a proper description 😄

File: piq/functional/base.py [hunk 4]
```diff
@@ -54,3 +54,28 @@ def gradient_map(x: torch.Tensor, kernels: torch.Tensor) -> torch.Tensor:
...
+    Returns:
+        Complex tensor with shape (2 x B x C x H x W)
```
Comment (snk4tr) [inline piq/functional/base.py:69]: And here about shapes again

File: piq/functional/colour_conversion.py [hunk 5]
```diff
@@ -113,3 +113,23 @@ def rgb2yiq(x: torch.Tensor) -> torch.Tensor:
...
+
+
+def rgb2lhm(x: torch.Tensor) -> torch.Tensor:
+    r"""Convert a batch of RGB images to a batch of LHM images
+
+    Args:
+        x: Batch of 4D (N x 3 x H x W) images in RGB colour space.
+
+    Returns:
+        Batch of 4D (N x 3 x H x W) images in LHM colour space.
```
Comment (snk4tr) [inline piq/functional/colour_conversion.py:125]: Previously you used `B` to denote batch size, but here `N` is used. It would be good to use the same letter for consistency

File: piq/functional/base.py (hunk 2 above)
Comment (denproc) [reply piq/functional/base.py:66]: No, the tensors have (2 B C H W) shapes in current implementation.

File: piq/functional/colour_conversion.py (hunk 5 above)
Comment (denproc) [reply piq/functional/colour_conversion.py:125]: However, all functions at `colour conversion` use `N`. I suggest to leave this to separate PR to unify the notation among all files of the library.

**Comment by zakajd** [comment]
> Here is a general suggestion: since we already have several unique Python metrics implementations (previously only MATLAB), it is a great advantage of our library comparing with similar ones. What do you @denproc @zakajd think about adding explicit notes about that somewhere on top of the README?

It's not always true, for example there was already [Python version of MDSI](https://github.com/igv/Python-MDSI). And same with some other metrics, like [BRISQUE](https://github.com/bukalapak/pybrisque),  [VSI](https://github.com/dingkeyan93/IQA-optimization/blob/master/IQA_pytorch/VSI.py), [FSIM](https://github.com/dingkeyan93/IQA-optimization/blob/master/IQA_pytorch/FSIM.py) and VIF [[1](https://github.com/dingkeyan93/IQA-optimization/blob/master/IQA_pytorch/VIF.py) [2](https://github.com/andrewekhalel/sewar/blob/master/sewar/full_ref.py)]. Our advantage is that we support GPU computation, gradients backpropagation etc. 

It seems that really unique are only GMSD and MS-GMSD metrics. In fact something called GMSD do [exist](https://github.com/dingkeyan93/IQA-optimization/blob/master/IQA_pytorch/GMSD.py), but computation is wrong and it should be called Gradient Similarity Measure (GSM)

File: piq/functional/base.py (hunk 2 above)
Comment (denproc) [reply piq/functional/base.py:66]: I tried to change the implementation to `(B x C x H x W x 2)`. It leads to reduction in precision and `nan` in gradients.

Comment (snk4tr) [reply piq/functional/base.py:66]: Then it is a bit weird shape, especially considering that:
1.  For SSIM/MS-SSIM implementations we have complex channels last https://github.com/photosynthesis-team/piq/blob/d3a0540fa2ea80724db11c2b5ecca08a80a7883e/piq/ssim.py#L25
2. It is kind of intuitive to have them last because we go from top to bottom: batches include channels, which consist of height and width pixels, which may be complex i.e. have two values: for real and imaginary parts respectively

@denproc would you consider re-implementing it with complex channels being last?

**Comment by snk4tr** [comment]
@zakajd my main idea was to clearly identify why would anyone bother installing and using our library. All things you listed here (including unique Python implementation of GMSD and MS-GMSD) are perfect examples of such competitive advantages. I will add an issue proposing this since it is important.

**Comment by zakajd** [comment]
@denproc  is it ready to be merged?
What about issue with complex tensors mentioned by Sergey? 

I agree with him that having complex dimension first is not a good idea. In PyTorch complex dimension is supposed to be the last one (for correct work of FFT and other complex operations). As far as I remember we agreed on similar order when you implemented #55

**Comment by denproc** [comment]
@zakajd @snk4tr I looked into it. Making the complex dimension last lead to reduction in precision and `nan` in gradients. Will take some time.

Regarding the description, I would suggest to refactor making description sell the library. It's becoming critical as the library is poorly indexed by search engines. 
One way to improve is creating a more concise list of measures (table with title and link).

**Comment by denproc** [comment]
@zakajd @snk4tr 
The problem with precision was located and led to calculation of the mean values (`mct_complex`). The problem was solved by calculating mean value in two steps independently.
The new `pow_for_complex` calculated with the case precision.
Overall, the precision comparing with MATLAB implementation remains the same.
